sql/src/main/java/io/crate/planner/operators/HashJoin.java

```diff
@@ -107,18 +103,38 @@ public ExecutionPlan build(PlannerContext plannerContext,
                                Row params,
                                Map<SelectSymbol, Object> subQueryValues) {
 
-        ExecutionPlan left = lhs.build(
+
+        ExecutionPlan leftExecutionPlan = lhs.build(
             plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
-        ExecutionPlan right = rhs.build(
+        ExecutionPlan rightExecutionPlan = rhs.build(
             plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
 
-        ResultDescription leftResultDesc = left.resultDescription();
-        ResultDescription rightResultDesc = right.resultDescription();
+        LogicalPlan leftLogicalPlan = lhs;
+        LogicalPlan rightLogicalPlan = rhs;
+
+        boolean tablesSwitched = false;
+        // Move smaller table to the right side
+        if (lhs.numExpectedRows() < rhs.numExpectedRows()) {
+            tablesSwitched = true;
+            leftLogicalPlan = rhs;
+            rightLogicalPlan = lhs;
+
+            ExecutionPlan tmp = leftExecutionPlan;
+            leftExecutionPlan = rightExecutionPlan;
+            rightExecutionPlan = tmp;
+        }
+
+        this.outputs.clear();
+        this.outputs.addAll(leftLogicalPlan.outputs());
+        this.outputs.addAll(rightLogicalPlan.outputs());
```

Comment:
I think its dangerous to mutate the outputs (they could be referenced by other operators already). whats wrong with the idea of using the EvalProjector to swap the outputs?

Reply:
You're right, I did this as a solution to the problems I encounter with the EvalProjection, but it's wrong...
Pushed a fixup where the EvalProjection solution works correctly now.